### PR TITLE
M1I13: WebSocketFrame

### LINKED ISSUE #13: M1I11: WebSocketFrame
## What it builds

`src/Server/Protocol/WebSocketFrame.php` — RFC 6455 frame encoder/decoder:
- `encode(string $data, int $opcode = 0x1): string` — payload → raw frame bytes
- `decode(string $data): ?self` — raw bytes → frame or `null` if incomplete
- `frameSize(string $data): ?int` — total frame byte size from header, or `null` if incomplete

Handles:
- Payload lengths: ≤125 (7-bit), 126 (16-bit extended), 127 (64-bit extended)
- Masking XOR for client→server frames
- Returns `null` when available bytes < required size (no silent truncation)

## Files to create

| Action | Path |
|--------|------|
| Create | `src/Server/Protocol/WebSocketFrame.php` |

## Dependencies

None (pure byte manipulation, no framework types).

## References

- Spec: WebSocketFrame section (part of WebSocketHandler)
- Plan: Task 12 (part 1)
- Watpliwosci: #24, #33
- Milestone doc: `docs/superpowers/milestone-1/11-websocket-frame.md`

### PR DESCRIPTION
**Pull request by s2x**
## Summary

| Revision | Changes |
|----------|---------|
| 1 | Initial implementation of RFC 6455 WebSocket frame encoder/decoder |

- Added `WebSocketFrame` value object with `encode()`, `decode()`, and `frameSize()` methods
- Handles all payload length variants (7-bit, 16-bit, 64-bit)
- Supports masking/unmasking for client→server frames
- Returns `null` for incomplete frames (no silent truncation)
- 23 unit tests covering encode, decode, frameSize, masking, extended lengths

## Acceptance criteria

- [x] `encode()` produces valid RFC 6455 frames with correct length encoding
- [x] `decode()` returns `null` for data shorter than required frame size (incomplete)
- [x] `decode()` returns `WebSocketFrame` for complete frames with correct payload
- [x] `frameSize()` reports total frame size including header
- [x] Masked payloads correctly XOR-unmasked
- [x] 16-bit and 64-bit extended payload lengths handled

Closes #13

### CODE REVIEW

File: src/Server/Protocol/WebSocketFrame.php [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    {
+        $unpacked = \unpack('J', substr($data, $offset, 8));
+
+        if ($unpacked === false) {
+            return null;
+        }
+
+        $value = $unpacked[1];
+
+        return \is_int($value) ? $value : null;
```
Comment (s2x) [inline src/Server/Protocol/WebSocketFrame.php]: **RESOLVED** — Fixed in 5f8e12c: added  guard to readUint64 and 2 regression tests.

File: src/Server/Protocol/WebSocketFrame.php [hunk 2]
```diff
@@ -0,0 +1,175 @@
...
+
+        $headerSize = 2;
+        $payloadLength = $payloadLengthCode;
+
+        if ($payloadLengthCode === 126) {
+            if ($len < $headerSize + 2) {
+                return null;
+            }
+            $read = self::readUint16($data, 2);
+            if ($read === null) {
```
Comment (s2x) [inline src/Server/Protocol/WebSocketFrame.php:149]: **RESOLVED** — Fixed in 5f8e12c: replaced literal  with tracked  variable in frameSize(), consistent with decode().

**Review by s2x — COMMENTED** [review]
## Review Summary — Round 1

| # | Severity | File:Line | Finding |
|---|----------|-----------|---------|
| 1 | MEDIUM | `src/Server/Protocol/WebSocketFrame.php:59` | `readUint64` may return negative int for 64-bit extended length with MSB set |
| 2 | LOW | `src/Server/Protocol/WebSocketFrame.php:148` | Magic number `2` for offset in `frameSize()` |

### Follow-up Observations
- Consider adding **opcode validation** in `decode()` — RFC 6455 reserves opcodes 0x3–0x7 and 0xB–0xF. Rejecting them would catch malformed frames early. File separately.
- `encode()` is fragmentation-unaware (always sets FIN=1, no `$fin` parameter). Acceptable for single-frame messages but worth planning for multi-frame support.
- XOR unmasking uses byte-by-byte `.=` concatenation — consider a more efficient approach (e.g. `str_repeat`‑based) if large WebSocket frames become common.
- `readUint64` returns `null` for 64‑bit values > PHP_INT_MAX on 32‑bit PHP (returns float, fails `is_int`). Unlikely to matter for a server‑side tool, but worth a comment.
- Missing PHPDoc blocks on the class and public methods (opcode values, RFC 6455 references).

### What's Fine :heavy_check_mark:
- All 6 acceptance criteria from the milestone doc are satisfied with test coverage
- `encode()`/`decode()`/`frameSize()` correctly handle all three payload length variants (≤125, 16‑bit, 64‑bit)
- Masked payloads correctly XOR‑unmasked
- Incomplete frames correctly return `null` at every boundary (empty, 1‑byte, short payload, short extended length field)
- Implementation matches the milestone plan — single file created, no unexpected side effects
- 23 unit tests with good coverage of edge cases

**Review by s2x — COMMENTED** [review]
## Review Summary — Round 2

All findings from Round 1 resolved. No new issues found.

| # | Severity | Finding | Status |
|---|----------|---------|--------|
| 1 | MEDIUM | `readUint64` negative overflow via MSB | **FIXED** — `$value >= 0` guard + 2 regression tests |
| 2 | LOW | Literal `2` offset in `frameSize()` | **FIXED** — tracked `$offset` variable, consistent with `decode()` |

### Follow-up Observations

- **Opcode validation**: The class accepts any 4-bit opcode (0x0–0xF). RFC 6455 only defines 0x0, 0x1, 0x2, 0x8, 0x9, 0xA. Reserved opcodes (0x3–0x7, 0xB–0xF) could be rejected in a future iteration.
- **`encode()` FIN fragmentation**: `encode()` always sets FIN=1. No continuation frames. By design per current spec.
- **XOR unmasking perf**: Byte-by-byte loop could be optimized with bulk string XOR for large payloads. Separate issue candidate.

### What's Fine

- All 6 acceptance criteria satisfied
- 25 tests covering all paths
- 7/16/64-bit payload lengths handled correctly
- Incomplete frame detection returns `null` at every boundary
- Masking XOR logic correct
- PHP 8.2 compat, PHPStan clean, CI green
